scenes/game/ai_player.gd

```diff
@@ -1,26 +1,82 @@
+## which is called by game.gd's `ai_take_turn`. `take_turn` consults the stored
+## game state the game state and returns an instance of one of the various
+## *Action classes defined in this file. `ai_take_turn` then uses that return
+## value to manifest the appropriate changes in the game proper.
+
+## `take_turn` enumerates all legal actions for the AI to take, then passes them
+## to policy (and then forwards the return value back to game.gd as described
+## above). Note that each possible variant of a (game) action counts as a
+## distinct (AI) action; for example, the choice of which cards to discard for
+## Force generation during a Walk.
+
+## A broad overview of the main chunks of this file:
+
+##   - class *State: Data wrappers for parts of game state.
+##   - class *Action: Represents an action the AI may or chooses to take,
+##         including parameters like which cards to discard in payment.
+##   - func get_*_actions: List all possible Action objects representing
+##         the named game action.
+##   - func pay_*, pick_*: Entry points for AI decisions that are not at
+##         the top of a turn; for example, a mid-strike decision on movement.
+
+class_name AIPlayer
+extends Resource
 
 const TEST_PrepareOnly = false
 
 const LocalGame = preload("res://scenes/game/local_game.gd")
-const AIPolicyRandom = preload("res://scenes/game/ai/ai_policy_random.gd")
-const AIPolicyRules = preload("res://scenes/game/ai/ai_policy_rules.gd")
 const Enums = preload("res://scenes/game/enums.gd")
 const CardDatabase = preload("res://scenes/game/card_database.gd")
 
-var game_player : LocalGame.Player
-var game_state : AIGameState = AIGameState.new()
-
-var ai_policy = AIPolicyRules.new()
+var game_logic : LocalGame
+var game_state : AIGameState
+var ai_policy
+
+# TODO: Check if unused in production? Seems like this should just be
+# encapsulated in game_state.
+var game_player: LocalGame.Player
+var game_opponent: LocalGame.Player
+
+func _init(local_game: LocalGame, player: LocalGame.Player, policy = null):
+	game_logic = local_game
+	game_player = player
+	game_opponent = local_game._get_player(local_game.get_other_player(player.my_id))
+	game_state = AIGameState.new(game_logic, game_player, game_opponent)
+	if policy == null:
+		ai_policy = AIPolicyRules.new()
+	else:
+		ai_policy = policy
 
 func set_ai_policy(new_policy):
+	ai_policy.free()
 	ai_policy = new_policy
 
-class AIPlayerState:
+## The AI states are static representations of game state; perhaps even the
+## current one. They replicate a bunch of primitives and basic collections so
+## that they can be reasoned about without affecting the state of the actual
+## game.
+
+class AIPlayerState extends Resource:
+	## The underlying game object that this player state reflects.
+	var source: LocalGame.Player
 	var player_id : Enums.PlayerId
-	var deck_id
 	var life
 	var deck
+	# `kit` is the character-defining JSON, which contains the UA/XA and
```

Comment:
let's just use deck_json then instead of defining a new term?

Reply:
or actually, I think I use deck_def inside local_game?

Reply:
Yeah, it's `deck_def`. Fixed.